janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphPSerializer.java

```diff
@@ -0,0 +1,128 @@
+// Copyright 2017 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.apache.tinkerpop.gremlin.process.traversal.P;
+import org.apache.tinkerpop.gremlin.process.traversal.util.AndP;
+import org.apache.tinkerpop.gremlin.process.traversal.util.ConnectiveP;
+import org.apache.tinkerpop.gremlin.process.traversal.util.OrP;
+import org.apache.tinkerpop.shaded.kryo.Kryo;
+import org.apache.tinkerpop.shaded.kryo.Serializer;
+import org.apache.tinkerpop.shaded.kryo.io.Input;
+import org.apache.tinkerpop.shaded.kryo.io.Output;
+import org.janusgraph.core.attribute.Geo;
+import org.janusgraph.core.attribute.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Serialize Janus Graph Predicates.
+ * 
+ * Since all predicates end up being an instance of P and the TinkerPop P
+ * serializer is not extensible, it is copied here and the JanusGraph predicates
+ * are added.
+ *
+ */
+public class JanusGraphPSerializer extends Serializer<P> {
```

Comment:
Is it possible instead to implement SerializerShim here for consistency with PSerializer? If so maybe you can pass through to the PSerializer write method to reduce some of the code redundancy. But I'm not sure the same would be possible for read() so I'm not sure it's worth it. Your call.

```java
public class JanusGraphPSerializer implements SerializerShim<P> {

    private GryoSerializers.PSerializer delegate = new GryoSerializers.PSerializer();

    @Override
    public <O extends OutputShim> void write(final KryoShim<?, O> kryo, final O output, final P p) {
        delegate.write(kryo, output, p);
    }
    ...
```